File: py/vcf/population.py
from collections import defaultdict
from typing import DefaultDict, Dict, Iterable, List, TextIO, Tuple


class Population:
    """
    This class represents a population, which is a list of individuals along
    with the group they belong to
    """
    
    def __init__(self) -> None:
        self.individual_to_group: Dict[str, str] = {}
        self.group_to_individuals: DefaultDict[str, List[str]] = defaultdict(list)
    
    
    def add_individual(self, identifier: str, group: str) -> None:
        """
        Add a new individual associated with the given group. Individual
        identifiers must be unique and can only belong to one group.
        """
        
        self.individual_to_group[identifier] = group
        self.group_to_individuals[group].append(identifier)
    
    
    def remove_individual(self, identifier: str) -> None:
        if identifier in self.individual_to_group:
            group = self.individual_to_group[identifier]
            del self.individual_to_group[identifier]
            self.group_to_individuals[group].remove(identifier)
    
    
    def individuals(self) -> Iterable[str]:
        """
        Returns an interator of the individuals in this population
        """
        
        return iter(self.individual_to_group)
    
    
    def groups(self) -> Iterable[str]:
        """
        Returns an interator of the groups in this population
        """
        
        return iter(self.group_to_individuals)
    
    
    def items(self) -> Iterable[Tuple[str, List[str]]]:
        """
        Return an item-like iterator that yields the various groups in this
        population. Each item is of the form
            (group_name, list_of_individuals)
        """
        
        return self.group_to_individuals.items()
    
    
    def has_individual(self, individual: str) -> bool:
        """
        Determines whether this population contains a specific individual.
        """
        
        return individual in self.individual_to_group


def parse_population(stream: TextIO) -> Population:
    """
    Read a text stream, where each non-empty line has two space-separated
    fields: an individual identifier and the group they belong to.
    """
    
    result = Population()
    
    for line in stream:
        assert isinstance(line, str)
        
        line = line.rstrip('\n')
        
        # Remove possible comments
        comment_start = line.find('#')
        if comment_start >= 0:
            line = line[:comment_start]
        line = line.strip()
        
        if not line:
            # Ignore empty lines
            continue
        
        fields = line.split()
        if len(fields) != 2:
            raise Exception(
                f'Line {line!r} is invalid: needs 2 fields, found {len(fields)}'
            )
        
        identifier, group = fields
        result.add_individual(identifier, group)
    
    return result

File: rs/src/vcf/population.rs
use std::collections::hash_map::Entry;
use std::collections::HashMap;
use std::io::{BufRead, Result, Error, ErrorKind};

pub struct Population {
    individual_to_group: HashMap<String, String>,
    group_to_individuals: HashMap<String, Vec<String>>,
}

impl Population {
    pub fn new() -> Self {
        Population {
            individual_to_group: HashMap::new(),
            group_to_individuals: HashMap::new(),
        }
    }

    pub fn add_individual(&mut self, individual: String, group: String) {
        self.individual_to_group.insert(individual, group);
    }

    pub fn group<Istr>(&self, individual: Istr) -> &str
    where
        Istr: AsRef<str>,
    {
        self.individual_to_group
            .get(individual.as_ref())
            .map(|s| &**s)
            .unwrap_or("???")
    }

    pub fn remove_individual<Istr>(&mut self, individual: Istr)
    where
        Istr: Into<String>,
    {
        if let Entry::Occupied(entry) = self.individual_to_group.entry(individual.into()) {
            let group = entry.remove();
            self.group_to_individuals.remove(&group);
        }
    }

    pub fn individuals(&mut self) -> impl Iterator<Item = &str> {
        self.individual_to_group.keys().map(|s| &**s)
    }

    pub fn groups(&mut self) -> impl Iterator<Item = &str> {
        self.group_to_individuals.keys().map(|s| &**s)
    }

    pub fn items(&mut self) -> impl Iterator<Item = (&str, &str)> {
        self.individual_to_group.iter().map(|(s1, s2)| (&**s1, &**s2))
    }

    pub fn has_individual<Istr>(&self, individual: Istr) -> bool
    where
        Istr: AsRef<str>,
    {
        self.individual_to_group.contains_key(individual.as_ref())
    }

    pub fn labels<Istr, I>(&self, individuals: I) -> impl Iterator<Item = &str>
    where
        I: IntoIterator<Item = Istr>,
        Istr: AsRef<str>,
    {
        individuals.into_iter().map(move |ind| self.group(ind))
    }
    
    pub fn parse<R: BufRead>(reader: R) -> Result<Population> {
        let mut population = Population::new();
        
        for line in reader.lines() {
            let line = line?;
            let line = line.trim_right_matches('\n');
            
            // Remove possible comments
            let line = if let Some(comment_start) = line.find('#') {
                &line[comment_start..]
            } else {
                line
            };
            
            // Remove extra heading and trailing spaces
            let line = line.trim();
            
            // Ignore empty lines
            if line.len() == 0 {
                continue;
            }
            
            let fields = line.split_whitespace().collect::<Vec<_>>();
            if fields.len() != 2 {
                let msg = format!("Line {:?} is invalid: needs 2 fields, found {}", line, fields.len());
                return Err(Error::new(ErrorKind::InvalidData, msg))
            }
            
            let identifier = fields[0];
            let group = fields[1];
            population.add_individual(identifier.to_string(), group.to_string());
        }
        
        Ok(population)
    }
}

